### PR TITLE
Versioning 1.0

### PR DESCRIPTION
**Pull request by strbean**
Problematic merge of versioning

### CODE REVIEW

File: website/django/data/migrations/0007_versioned_data.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    dependencies = [
+        ('data', '0006_versioned_data'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='variant',
+            name='Data_Release',
+            # Get data_release_id_seq.start_value instead of assuming 1 ?
```
Comment (zfisch) [inline website/django/data/migrations/0007_versioned_data.py]: Following up on this, it looks like this migration will fail. This is the error I received when running the migration locally:
```
(brca) Zacharys-MacBook-Air:django zackfischmann$ python manage.py migrate
/Users/zackfischmann/UCSC/brca-exchange/website/django/brca/urls.py:29: RemovedInDjango110Warning: Support for string view arguments to url() is deprecated and will be removed in Django 1.10 (got django.views.static.serve). Pass the callable instead.
  {'document_root': settings.MEDIA_ROOT, 'show_indexes': True})

/Users/zackfischmann/UCSC/brca-exchange/website/django/brca/urls.py:29: RemovedInDjango110Warning: django.conf.urls.patterns() is deprecated and will be removed in Django 1.10. Update your urlpatterns to be a list of django.conf.urls.url() instances instead.
  {'document_root': settings.MEDIA_ROOT, 'show_indexes': True})

/Users/zackfischmann/UCSC/brca-exchange/website/django/brca/urls.py:34: RemovedInDjango110Warning: Support for string view arguments to url() is deprecated and will be removed in Django 1.10 (got django.views.static.serve). Pass the callable instead.
  {'document_root': settings.STATIC_ROOT, 'show_indexes': True})

/Users/zackfischmann/UCSC/brca-exchange/website/django/brca/urls.py:34: RemovedInDjango110Warning: django.conf.urls.patterns() is deprecated and will be removed in Django 1.10. Update your urlpatterns to be a list of django.conf.urls.url() instances instead.
  {'document_root': settings.STATIC_ROOT, 'show_indexes': True})

Operations to perform:
  Apply all migrations: authtoken, users, sessions, admin, sites, auth, contenttypes, data
Running migrations:
  Rendering model states... DONE
  Applying data.0007_versioned_data...Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 200, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/migrations/executor.py", line 92, in migrate
    self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/migrations/executor.py", line 121, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/migrations/executor.py", line 198, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 90, in __exit__
    self.execute(sql)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 110, in execute
    cursor.execute(sql, params)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/Users/zackfischmann/.virtualenvs/brca/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: insert or update on table "variant" violates foreign key constraint "variant_Data_Release_id_34907b0c_fk_data_release_id"
DETAIL:  Key (Data_Release_id)=(1) is not present in table "data_release".
```

**Comment by david4096** [comment]
Is this currently running at http://brcaexchange-dev.cloudapp.net:9003 ? Is there a beta site to test against?

**Comment by david4096** [comment]
I wanted to pass along a first round of testing:

I was not able to apply migrations successfully.

Tried starting from scratch with `python manage.py migrate --fake data zero && python manage.py migrate` and got `django.db.utils.ProgrammingError: relation "data_release" already exists`.

Operations to perform:
  Apply all migrations: authtoken, users, sessions, admin, sites, auth, contenttypes, data
Running migrations:
  Rendering model states... DONE
  Applying data.0001_initial... OK
  Applying data.0002_search_index... OK
  Applying data.0004_autocomplete_words... OK
  Applying data.0006_versioned_data...Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/commands/migrate.py", line 200, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/migrations/executor.py", line 92, in migrate
    self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/migrations/executor.py", line 121, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/migrations/executor.py", line 198, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/migrations/migration.py", line 123, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/migrations/operations/models.py", line 59, in database_forwards
    schema_editor.create_model(model)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 284, in create_model
    self.execute(sql, params or None)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/backends/base/schema.py", line 110, in execute
    cursor.execute(sql, params)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/db/backends/utils.py", line 62, in execute
    return self.cursor.execute(sql)
django.db.utils.ProgrammingError: relation "data_release" already exists

I wiped the tables using:

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;
ALTER SCHEMA public OWNER to postgres;
```

I then found I was able to run migrations but the step that added the data was removed. I'm now taking a look at redoing the migrations.

**Comment by david4096** [comment]
I've found the addrelease script but wasn't sure where the metadata JSON (notes) can be found so I wrote this one `{"sources": [], "id": "9", "date": "2016-10-26 21:46:06"}`.

The problem I have now is:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 353, in execute_from_command_line
    utility.execute()
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/__init__.py", line 345, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/base.py", line 348, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/david/brca-exchange/website/django/env/lib/python2.7/site-packages/django/core/management/base.py", line 399, in execute
    output = self.handle(*args, **options)
  File "/Users/david/brca-exchange/website/django/data/management/commands/addrelease.py", line 39, in handle
    row_dict['Genomic_Coordinate_hg38'] = row_dict.pop('pyhgvs_Genomic_Coordinate_38')
KeyError: 'pyhgvs_Genomic_Coordinate_38'
```

Which suggests to me that the `aggregated.tsv` in the source tree has been outdated?

How does one generate the test database?

**Comment by zfisch** [comment]
Hi @david4096 there's definitely several issues with migrations. I'm pushing up some major fixes today after a little more local testing.

Regarding aggregated.tsv -- this will be removed and replaced with new data that has a different schema. A version.json file (metadata) will be included with the data for each release (every time we generate new data, a new version.json file will be included that contains information about the release). I'll include all of this in my next commits as well.

This code is not currently up on a dev environment (I know @strbean had an updated version of it at 9002, but I believe he's testing further changes now that diverge from this branch).

**Comment by david4096** [comment]
Great, thanks @zfisch ! Ping me when you are confident there is a build process I can work with. I wasn't able to get our tests to pass.

**Comment by zfisch** [comment]
@david4096 I just cleaned up some of the test suite but am noticing failures in the GA4GH tests.

In order to get your test environment set up properly, you'll want to pull the latest versioning-1.0 branch, run `./manage.py migrate --fake data zero` to reset your migrations, then run `./manage.py migrate data` to apply the new migrations and pick up the new schema (NOTE: this will drop your variant, data_release, words, and data_changetype tables and replace them).

**Comment by david4096** [comment]
@zfisch I was able to run the new migrations from scratch without error, can you point me to a place for the data?

**Comment by david4096** [comment]
Great, thanks! Looks like there is one failure and two errors.  Would I push straight to this branch?

**Comment by david4096** [comment]
I was able to get two of the three tests to pass [here](https://github.com/david4096/brca-exchange/tree/versioning-1.0). The failing test has to do with how GA4GH counts. 

We were relying on parsing the genomic coordinate, and it seems that for the hg38 mapped variants, the hgvs coordinate is the same as what is in the `hg38_Start` column. We need to add a test that guarantees for some specific variant we've gotten the coordinates right in each reference set.

**Comment by zfisch** [comment]
@david4096 I think it makes the most sense to push to this branch until it is merged into master since most GA4GH changes are already included (TBH I'm not clear as to whether the GA4GH work was supposed to be merged in to this versioning-1.0 branch, but I'm tasked with getting this branch fixed up and merged into master ASAP).

@melissacline could you speak to @david4096's question about hg38 counting from 0?

**Comment by david4096** [comment]
We'll get it working! @dkucsc noted that we were incorrectly returning a 404 when we meant to return a 400, and that at times we returned a 400 mistakenly so the extra round of testing so we'll slip a bug fix in there. Will this code be staged somewhere?

```
def latest_variants():
    """Helper function provides variants from the latest release"""
    latest = DataRelease.objects.order_by('-id')[0].id
    variants = Variant.objects.filter(Data_Release_id = latest)
    return variants
```

Please note how I chose to find the latest variants as part of our search requests. Are there any other filters I have to apply to make sure there aren't duplicates? We could consider having a variant set per data release.

I don't have push rights on this repository, but I'm currently updated with your tree [here](https://github.com/david4096/brca-exchange/tree/versioning-1.0).

**Comment by zfisch** [comment]
@david4096 awesome! This code will be on the staging site (http://brcaexchange.cloudapp.net/) after merging into master. It will not be live on the production site until we manually start the deploy to master process which will only happen after confirming everything is ready on staging.

From what I understand, the code you shared should only return the latest variants, though I believe you could also use the currentvariant object -- @strbean can you confirm this?

**Comment by david4096** [comment]
So it seems like we have a working branch now! We can move ahead and I'll leave my question open regarding Hg38 for now. Looking forward to further testing!

**Comment by zfisch** [comment]
Wonderful! Before merging I'm going to set this up on my dev box and share with everybody just to be sure. Should have it up by tomorrow afternoon, will keep everybody posted.

**Pull request merged**